keras_cv/layers/object_detection/iou_similarity.py

```diff
@@ -0,0 +1,103 @@
+    are unbatched and by [`batch`, `boxes1_index`,`boxes2_index`] if the boxes are
+    batched.
+
+    Users can configure to mask the result. `box1_mask` will mask `boxes1` and `box2_mask`
+    will mask `boxes2`, the result will be masked by `mask_val`. Boxes are masked when they're
+    out of the boundary, i.e., all their coodinates are less than 0. Users can pad ground truth
+    boxes with -1. in order for it to be masked here.
+
+    Args:
+      boxes1: a list of bounding boxes in 'corners' format. Can be batched or unbatched.
+      boxes2: a list of bounding boxes in 'corners' format. This should match the rank and
+        shape of boxes1.
+      bounding_box_format: a case-insensitive string.
+        For detailed information on the supported format, see the
+        [KerasCV bounding box documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
+
+    Returns:
+      iou_lookup_table: a vector containing the pairwise ious of boxes1 and
+        boxes2.
+    """
+
+    def __init__(
+        self,
+        box1_format,
+        box2_format,
+        box1_mask=False,
+        box2_mask=False,
+        mask_val=-1,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.box1_format = box1_format
+        self.box2_format = box2_format
+        self.mask_val = mask_val
+        self.box1_mask = box1_mask
+        self.box2_mask = box2_mask
+        self.built = True
+
+    def call(self, boxes1, boxes2, boxes1_mask=None, boxes2_mask=None):
+        boxes1 = bounding_box.convert_format(
+            boxes1, source=self.box1_format, target="xyxy"
+        )
+        boxes2 = bounding_box.convert_format(
+            boxes2, source=self.box2_format, target="xyxy"
+        )
+        result = bounding_box.compute_iou(boxes1, boxes2, bounding_box_format="xyxy")
+        if not self.box1_mask and not self.box2_mask:
+            return result
+        background_mask = None
+        boxes2_rank = len(boxes2.shape)
+        mask_val_t = tf.cast(self.mask_val, result.dtype) * tf.ones_like(result)
+        perm = [1, 0] if boxes2_rank == 2 else [0, 2, 1]
+        if self.box1_mask and self.box2_mask:
+            boxes1_mask = tf.less(tf.reduce_max(boxes1, axis=-1, keepdims=True), 0.0)
+            boxes2_mask = tf.less(tf.reduce_max(boxes2, axis=-1, keepdims=True), 0.0)
+            background_mask = tf.logical_or(
+                boxes1_mask, tf.transpose(boxes2_mask, perm)
+            )
+        elif self.box1_mask:
+            boxes1_mask = tf.less(tf.reduce_max(boxes1, axis=-1, keepdims=True), 0.0)
+            background_mask = boxes1_mask
+        else:
+            boxes2_mask = tf.less(tf.reduce_max(boxes2, axis=-1, keepdims=True), 0.0)
+            background_mask = tf.logical_or(
+                tf.zeros(tf.shape(boxes2)[:-1], dtype=tf.bool),
+                tf.transpose(boxes2_mask, perm),
+            )
+        iou_lookup_table = tf.where(background_mask, mask_val_t, result)
```

Comment:
Any reason for this to be a layer instead of a utility?  

I wouldn't expect this to be plugged anywhere else into the API in a configurable way (i.e. as a constructor parameter for another component), so I think it makes a lot of sense to keep it as a function.